core/sdk-core/src/main/java/software/amazon/awssdk/core/endpointdiscovery/EndpointDiscoveryRefreshCache.java

```diff
@@ -87,6 +90,64 @@ public URI get(String accessKey, EndpointDiscoveryRequest request) {
         return endpoint.endpoint();
     }
 
+    public CompletableFuture<URI> getAsync(String accessKey, EndpointDiscoveryRequest request) {
+        key = accessKey;
+
+        // Support null (anonymous credentials) by mapping to empty-string. The backing cache does not support null.
+        if (key == null) {
+            key = "";
+        }
+
+        if (request.cacheKey().isPresent()) {
+            key = key + ":" + request.cacheKey().get();
+        }
+
+        EndpointDiscoveryEndpoint endpoint = cache.get(key);
+
+        if (endpoint == null) {
+            if (request.required()) {
+                return discoverEndpoint(request).handle(
+                    (endpointDiscoveryEndpoint, throwable) -> {
+                        if (throwable != null) {
+                            if (throwable instanceof InterruptedException) {
+                                Thread.currentThread().interrupt();
+                                throw EndpointDiscoveryFailedException.create(throwable);
+                            }
+                            if (throwable instanceof ExecutionException
```

Comment:
It's not necessary to catch `InterruptedException`/or re-interrupt, especially since we have no idea what thread actually got interrupted; it may not be the same as `Thread.currentThread()`

Same for `ExecutionException` since we're not waiting on a future.get() anymore

Reply:
Ah right. We only need to catch the CompletableFuture exception in that case. I will change it